src/common/utilities/FormulaEvaluation.ts

```diff
@@ -0,0 +1,641 @@
+            [/^\w+/, "WORD"],                                   // Other words, checked against valid variables
+            [/^&&|^\|\||^==|^<>/, "OPERATOR"],                        // Operators and special characters (match double first)
+            [/^[+\-*/<>=%!&|?:,()[\]]/, "OPERATOR"],           // Operators and special characters
+        ];
+
+        const tokens: Token[] = [];
+        let i = 0;
+
+        while (i < expression.length) {
+            let match: string | null = null;
+
+            // For each pattern, try to match it from the current position in the expression
+            for (const [pattern, tokenType] of patterns) {
+                const regexResult = pattern.exec(expression.slice(i));
+
+                if (regexResult) {
+                    match = regexResult[1] || regexResult[0];
+
+                    // Unary minus is a special case that we need to 
+                    // capture in order to process negative numbers, or 
+                    // expressions such as 5 + - 3
+                    if (tokenType === "OPERATOR" && match === "-" && (tokens.length === 0 || tokens[tokens.length - 1].type === "OPERATOR")) {
+                        tokens.push(new Token("UNARY_MINUS", match));
+                        i += match.length + expression.slice(i).indexOf(match);
+                    } else if (
+                        // String literals, surrounded by single or double quotes
+                        tokenType === "STRING"
+                    ) {
+                        tokens.push(new Token(tokenType, match));
+                        i += match.length + 2;
+                    } else if (tokenType === "WORD") {
+                        // We only match words if they are a valid variable name
+                        if (context && context[match]) {
+                            tokens.push(new Token("VARIABLE", match));
+                        }
+                        i += match.length;
+                    } else {
+                        // Otherwise, just add the token to the list
+                        tokens.push(new Token(tokenType, match));
+                        // console.log(`Added token: ${JSON.stringify({tokenType: tokenType, value: match})}`);
+                        i += match.length + expression.slice(i).indexOf(match);
+                        if (tokenType === "FUNCTION") i -= 1;
+                    }
+
+                    break;
+                }
+            }
+
+            if (!match) {
+                // If no patterns matched, move to the next character
+                // console.log(`No match found for character: ${expression[i]}`);
+                i++;
+            }
+        }
+
+        return tokens;
+    }
+
+    public shuntingYard(tokens: Token[]): Token[] {
+
+        /** Returns a precedence value for a token or operator */
+        function getPrecedence(op: string): { precedence: number, associativity: "left" | "right" } {
```

Comment:
I don't think you need a closure here as you're not referencing parent's variables/args.
I'd suggest making it a regular private method of the class.

Reply:
This was a holdover from prototyping outside of the project. I've made it into a private method of the class as suggested 👍